providers/base/bin/smartcard_test.py

```diff
@@ -0,0 +1,329 @@
+                    and real_reader.name == card.reader
+                ):
+                    cards.append(card)
+                    self.logger.info("Smart card insertion detected:")
+                    self.logger.info(card)
+                    self.logger.info(
+                        "\nPlease remove it to test the removal detection\n"
+                    )
+                    break
+
+        cardrequest = CardRequest(timeout=30, newcardonly=False)
+        while True:
+            currentcards = cardrequest.waitforcardevent()
+            for card in cards:
+                if not currentcards.__contains__(card):
+                    cards.remove(card)
+                    self.logger.info("Smart card removal detected:")
+                    self.logger.info(card)
+                    return
+
+    def send_apdu_test(self, reader: str):
+        """
+        Send the APDU command to the smart card and verify the response.
+
+        :param reader: Stringified smart card reader name
+        """
+        sc_conn = self.get_connection(reader)
+        if sc_conn:
+            self.logger.info("ATR from smartcard:")
+            self.logger.info(toHexString(sc_conn.getATR()))
+            # This is a sample APDU command
+            SELECT = [0xA0, 0xA4, 0x00, 0x00, 0x02]
+            DF_TELECOM = [0x7F, 0x10]
+            data, sw1, sw2 = sc_conn.transmit(SELECT + DF_TELECOM)
+            if sw1 in self.sw1_list:
+                self.logger.info("Send/Receive APDU command is working")
+                return
+        raise SystemExit("Could not working for this smartcard reader")
+
+    def _args_parsing(self, args=sys.argv[1:]):
+        """
+        command line arguments parsing
+
+        :param args: arguments from sys
+        :type args: sys.argv
+        """
+        parser = argparse.ArgumentParser(
+            prog="Smartcard validator",
+            description="use to test smartcard reader could work",
+        )
+
+        subparsers = parser.add_subparsers(dest="test_type")
+        subparsers.required = True
+
+        # Add parser for listing smartcard readers for resource job
+        parser_resources = subparsers.add_parser(
+            "resources",
+            help="list smartcard readers on this system to resources",
+        )
+        parser_resources.add_argument(
+            "-t",
+            "--type",
+            type=str,
+            default="All",
+            help="""
+                  List All/contact/contactless smartcard reader for testing
+                  (default: %(default)s)
+                 """,
+        )
+
+        # Add parser for listing smartcard readers for resource job
+        parser_detect_reader = subparsers.add_parser(
+            "detect_reader",
+            help="count smartcard readers on this system",
+        )
+        parser_detect_reader.add_argument(
+            "-t",
+            "--type",
+            type=str,
+            default="All",
+            help="""
```

Comment:
might need to restrict input values here since there's no runtime checks:

```py
parser_detect_reader.add_argument(
    "-t",
    "--type",
    choices=['contact',contactless','all']
)
```

Reply:
Or save it as a global constant `READER_TYPES = ("contact", "contactless", "all")`